File: 16.最接近的三数之和.go
/*
 * @lc app=leetcode.cn id=16 lang=golang
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	length := len(nums)
	best := math.MaxInt32

	update := func(cur int) {
		if abs(cur-target) < abs(best-target) {
			best = cur
		}
	}

	for first := 0; first < length; first++ {
		if first > 0 && nums[first] == nums[first-1] {
			continue
		}

		second, third := first+1, length-1
		for second < third {
			sum := nums[first] + nums[second] + nums[third]
			// 如果直接相等，返回之
			if sum == target {
				return sum
			}

			update(sum)

			// 更新下标
			if sum > target {
				third0 := third - 1
				// 计算的至一定要大 second
				for second < third0 && nums[third] == nums[third0] {
					third0--
				}
				third = third0
			}
			if sum < target {
				second0 := second + 1
				// 计算的至一定要小third
				for second0 < third && nums[second] == nums[second0] {
					second0++
				}
				second = second0
			}

		}
	}

	return best
}

func abs(x int) int {
	if x < 0 {
		return -1 * x
	}
	return x
}

// @lc code=end

File: 16.最接近的三数之和.py
from typing import *
#
# @lc app=leetcode.cn id=16 lang=python3
#
# [16] 最接近的三数之和
#

# @lc code=start

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        length = len(nums)
        nums.sort()
        diff = 7**10

        # n**3 耗时太长。go用双指针方法
        for first in range(length):
            if first > 0 and nums[first] == nums[first - 1]:
                continue

            for second in range(first + 1, length):
                if second > first + 1 and nums[second] == nums[second - 1]:
                    continue
                # 这里因为abs函数不是单调的，不能用此方法，要不然会漏掉第一个小于diff后的 third
                # while second < third and abs(nums[first] + nums[second] + nums[third] - target) > diff:
                #     third -= 1
                # if second == third:
                #     break
                # if abs(nums[first] + nums[second] + nums[third] - target) < diff:
                #     diff = abs(nums[first] + nums[second] + nums[third] - target)
                #     best = nums[first] + nums[second] + nums[third]
                for third in range(second + 1, length):
                    if third > third + 1 and nums[third] == nums[third - 1]:
                        continue
                    if abs(nums[first] + nums[second] + nums[third] - target) < diff:
                        diff = abs(nums[first] + nums[second] + nums[third] - target)
                        best = nums[first] + nums[second] + nums[third] 
        
        return best



# @lc code=end

s = Solution()
print(s.threeSumClosest([1,1,-1,-1,3], -1))